src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -1020,6 +1023,158 @@ void MeshForwarder::UpdateRoutes(uint8_t *           aFrame,
+            (mFragmentEntries[i].GetSrcRloc16() == aSrcRloc16))
+        {
+            break;
+        }
+    }
+
+    return (i >= OT_ARRAY_LENGTH(mFragmentEntries)) ? NULL : &mFragmentEntries[i];
+}
+
+FragmentPriorityEntry *MeshForwarder::GetUnusedFragementPriorityEntry(void)
+{
+    size_t i;
+
+    for (i = 0; i < OT_ARRAY_LENGTH(mFragmentEntries); i++)
+    {
+        if (mFragmentEntries[i].GetLifetime() == 0)
+        {
+            break;
+        }
+    }
+
+    return (i >= OT_ARRAY_LENGTH(mFragmentEntries)) ? NULL : &mFragmentEntries[i];
+}
+
+otError MeshForwarder::GetFragmentPriority(Lowpan::FragmentHeader &aFragmentHeader,
+                                           uint16_t                aSrcRloc16,
+                                           uint8_t &               aPriority)
+{
+    otError                error = OT_ERROR_NONE;
+    FragmentPriorityEntry *entry;
+
+    VerifyOrExit((entry = FindFragmentPriorityEntry(aFragmentHeader.GetDatagramTag(), aSrcRloc16)) != NULL,
+                 error = OT_ERROR_NOT_FOUND);
+    aPriority = entry->GetPriority();
+
+exit:
+    return error;
+}
+
+otError MeshForwarder::GetForwardFramePriority(const uint8_t *     aFrame,
+                                               uint8_t             aFrameLength,
+                                               const Mac::Address &aMacDest,
+                                               const Mac::Address &aMacSource,
+                                               uint8_t &           aPriority)
+{
+    otError                error = OT_ERROR_NONE;
+    Lowpan::MeshHeader     meshHeader;
+    Lowpan::FragmentHeader fragmentHeader;
+
+    SuccessOrExit(error = GetMeshHeader(aFrame, aFrameLength, meshHeader));
+    aFrame += meshHeader.GetHeaderLength();
+    aFrameLength -= meshHeader.GetHeaderLength();
+
+    if (GetFragmentHeader(aFrame, aFrameLength, fragmentHeader) == OT_ERROR_NONE)
+    {
+        aFrame += fragmentHeader.GetHeaderLength();
+        aFrameLength -= fragmentHeader.GetHeaderLength();
+
+        if (fragmentHeader.GetDatagramOffset() == 0)
+        {
+            // Get priority from Ipv6 header or UDP destination port directly
+            SuccessOrExit(error = GetFramePriority(aFrame, aFrameLength, aMacSource, aMacDest, aPriority));
+        }
+        else
+        {
+            // Get priority from the pre-buffered info
+            SuccessOrExit(error = GetFragmentPriority(fragmentHeader, meshHeader.GetSource(), aPriority));
+        }
+
+        UpdateFragmentPriority(fragmentHeader, aFrameLength, meshHeader.GetSource(), aPriority);
+    }
+    else
+    {
+        // Get priority from Ipv6 header or UDP destination port directly
+        error = GetFramePriority(aFrame, aFrameLength, aMacSource, aMacDest, aPriority);
+    }
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        otLogWarnMac("Failed to get forwarded frame priority, error:%s, len:%d, dst:%s, src:%s",
```

Comment:
Suggest using `NOTICE` level here since this is not really an error condition - things will still operate normally, just not as optimally.

Reply:
Updated.